File: examples/gstreamer/lat.py
#!/usr/bin/env python2.2
#
# gst-python
# Copyright (C) 2002 David I. Lehn
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.
# 
# Author: David I. Lehn <dlehn@vt.edu>
#

import sys
import time
from gstreamer import *
from gobject import GObject

def update(sender, *args):
   print sender.get_name(), args

max = 0
min = -1
total = 0
count = 0
print_del = 1
interations = 0

def handoff_src(src, buf):
   #buf.set_timestamp(time.time())
   pass

def handoff_sink(sink, buf):
   global max, min, total, count

   end = time.time()
   #d = end - buf.get_timestamp()
   d = end - 0
   if d > max:
      max = d
   if d < min:
      min = d
   total += d
   count += 1
   avg = total/count

   if (count % print_del) == 0:
      print '%07d:%08d min:%08d max:%08d avg:%f\n' %\
            (count, d, min, max, avg),

def identity_add(pipeline, first, count):
   last = first

   for i in range(count):
     name = 'identity_%03d' % i
     ident = gst_elementfactory_make('identity', name)
     assert ident
     ident.set_property('silent', 1)
     pipeline.add(ident)
     last.get_pad('src').connect(ident.get_pad('sink'))
     last = ident

   return last

def fakesrc():
   src = gst_elementfactory_make('fakesrc','src')
   assert src
   src.set_property('silent', 1)
   src.set_property('num_buffers', iterations)
   GObject.connect(src, 'handoff', handoff_src)
   return src

def fakesink():
   sink = gst_elementfactory_make('fakesink','fakesink')
   assert sink
   sink.set_property('silent', 1)
   GObject.connect(sink, 'handoff', handoff_sink)
   return sink

def simple(argv):
   if len(argv) < 1:
     print 'simple: bad params'
     return None
   idents = int(argv[0])
   if len(argv) == 2:
      gst_schedulerfactory_set_default_name (argv[1])

   pipeline = gst_pipeline_new('pipeline')
   assert pipeline

   src = fakesrc()
   pipeline.add(src)
   last = identity_add(pipeline, src, idents)
   sink = fakesink()
   pipeline.add(sink)
   last.get_pad('src').connect(sink.get_pad('sink'))

   return pipeline

def queue(argv):
   if len(argv) < 1:
      print 'queue: bad params'
      return None
   idents = int(argv[0])

   if len(arv) == 2:
      gst_schedulerfactory_set_default_name (argv[1])

   pipeline = gst_pipeline_new('pipeline')
   assert pipeline

   src_thr = gst_thread_new('src_thread')
   assert src_thr

   src = fakesrc()
   assert src
   src_thr.add(src)

   src_q = gst_elementfactory_make('queue','src_q')
   assert src_q
   src_thr.add(src_q)
   src.get_pad('src').connect(src_q.get_pad('sink'))

   pipeline.add(src_thr)

   last = identity_add(pipeline, src_q, idents)

   sink_q = gst_elementfactory_make('queue','sink_q')
   assert sink_q
   pipeline.add(sink_q)
   last.get_pad('src').connect(sink_q.get_pad('sink'))

   sink_thr = gst_thread_new('sink_thread')
   assert sink_thr

   sink = fakesink()
   assert sink

   sink_thr.add(sink)

   pipeline.add(sink_thr)

   sink_q.get_pad('src').connect(sink.get_pad('sink'))

   return pipeline

tests = {
   'simple' : ('ident_count [scheduler_name]', simple),
   'queue' : ('ident_count [scheduler_name]', queue),
}

def main():
   "A GStreamer latency tester"
   #gst_debug_set_categories(-1)
   global iterations, print_del

   if len(sys.argv) < 3:
      print 'usage: %s iterations print_del test_name [test_params...]' % sys.argv[0]
      for name in tests.keys():
         doc, func = tests[name]
         print '  %s %s' % (name, doc)
      return -1
   else:
      iterations = int(sys.argv[1])
      print_del = int(sys.argv[2])
      name = sys.argv[3]
 
   pipeline = tests[name][1](sys.argv[4:])
   assert pipeline

   #xmlSaveFile('lat.gst', gst_xml_write(pipeline))

   pipeline.set_state(STATE_PLAYING)

   while count < iterations:
      pipeline.iterate()

   pipeline.set_state(STATE_NULL)

   print

   return 0;

if __name__ == '__main__':
   ret = main()
   sys.exit (ret)

File: tests/lat.c
#include <gst/gst.h>
#include <stdlib.h>
#include <string.h>

/* FIXME: WTF does this do? */

static guint64 max = 0, min = -1, total = 0;
static guint count = 0;
static guint print_del = 1;
static guint iterations = 0;
static guint mhz = 0;

void
handoff_src (GstElement * src, GstBuffer * buf, gpointer user_data)
{
  gst_trace_read_tsc (&GST_BUFFER_TIMESTAMP (buf));
}

void
handoff_sink (GstElement * sink, GstBuffer * buf, gpointer user_data)
{
  guint64 end, d, avg;
  guint avg_ns;

  gst_trace_read_tsc (&end);
  d = end - GST_BUFFER_TIMESTAMP (buf);
  if (d > max)
    max = d;
  if (d < min)
    min = d;
  total += d;
  count++;
  avg = total / count;
  avg_ns = (guint) (1000.0 * (double) avg / (double) mhz);

  if ((count % print_del) == 0) {
    g_print ("%07d:%08" G_GUINT64_FORMAT " min:%08" G_GUINT64_FORMAT " max:%08"
	G_GUINT64_FORMAT " avg:%08" G_GUINT64_FORMAT " avg-s:0.%09d\r", count,
	d, min, max, avg, avg_ns);
  }
}

GstElement *
identity_add (GstPipeline * pipeline, GstElement * first, int count)
{
  GstElement *last, *ident;
  int i;
  char buf[20];

  last = first;

  for (i = 0; i < count; i++) {
    snprintf (buf, 20, "identity_%03d", i);
    ident = gst_element_factory_make ("identity", buf);
    g_return_val_if_fail (ident != NULL, NULL);
    g_object_set (G_OBJECT (ident), "silent", TRUE, NULL);
    gst_bin_add (GST_BIN (pipeline), GST_ELEMENT (ident));
    gst_pad_link (gst_element_get_pad (last, "src"),
	gst_element_get_pad (ident, "sink"));
    last = ident;
  }

  return last;
}

GstElement *
fakesrc (void)
{
  GstElement *src;

  src = gst_element_factory_make ("fakesrc", "src");
  g_return_val_if_fail (src != NULL, NULL);
  g_object_set (G_OBJECT (src), "silent", TRUE, NULL);
  g_object_set (G_OBJECT (src), "num_buffers", iterations, NULL);
  g_signal_connect (G_OBJECT (src), "handoff", G_CALLBACK (handoff_src), NULL);

  return src;
}

GstElement *
fakesink (void)
{
  GstElement *sink;

  sink = gst_element_factory_make ("fakesink", "fakesink");
  g_return_val_if_fail (sink != NULL, NULL);
  g_object_set (G_OBJECT (sink), "silent", TRUE, NULL);
  g_signal_connect (G_OBJECT (sink),
      "handoff", G_CALLBACK (handoff_sink), NULL);

  return sink;
}

GstPipeline *
simple (int argc, int argi, char *argv[])
{
  GstPipeline *pipeline;
  GstElement *last, *src, *sink;
  int idents;

  if ((argc - argi) < 1) {
    fprintf (stderr, "bad params");
    return NULL;
  }
  idents = atoi (argv[argi]);
  if ((argc - argi) == 2) {
    gst_scheduler_factory_set_default_name (argv[argi + 1]);
  }

  pipeline = GST_PIPELINE (gst_pipeline_new ("pipeline"));
  g_return_val_if_fail (pipeline != NULL, NULL);

  src = fakesrc ();
  gst_bin_add (GST_BIN (pipeline), GST_ELEMENT (src));
  last = identity_add (pipeline, src, idents);
  sink = fakesink ();
  gst_bin_add (GST_BIN (pipeline), GST_ELEMENT (sink));
  gst_pad_link (gst_element_get_pad (last, "src"),
      gst_element_get_pad (sink, "sink"));

  return pipeline;
}

GstPipeline *
queue (int argc, int argi, char *argv[])
{
  GstPipeline *pipeline;
  GstElement *last, *src, *sink, *src_thr, *src_q, *sink_q, *sink_thr;
  int idents;

  if ((argc - argi) < 1) {
    fprintf (stderr, "bad params");
    return NULL;
  }
  idents = atoi (argv[argi]);

  if ((argc - argi) == 2) {
    gst_scheduler_factory_set_default_name (argv[argi + 1]);
  }

  pipeline = GST_PIPELINE (gst_pipeline_new ("pipeline"));
  g_return_val_if_fail (pipeline != NULL, NULL);

  src_thr = GST_ELEMENT (gst_thread_new ("src_thread"));
  g_return_val_if_fail (src_thr != NULL, NULL);

  src = fakesrc ();
  g_return_val_if_fail (src != NULL, NULL);
  gst_bin_add (GST_BIN (src_thr), GST_ELEMENT (src));

  src_q = gst_element_factory_make ("queue", "src_q");
  g_return_val_if_fail (src_q != NULL, NULL);
  gst_bin_add (GST_BIN (src_thr), GST_ELEMENT (src_q));
  gst_pad_link (gst_element_get_pad (src, "src"),
      gst_element_get_pad (src_q, "sink"));

  gst_bin_add (GST_BIN (pipeline), GST_ELEMENT (src_thr));

  last = identity_add (pipeline, src_q, idents);

  sink_q = gst_element_factory_make ("queue", "sink_q");
  g_return_val_if_fail (sink_q != NULL, NULL);
  gst_bin_add (GST_BIN (pipeline), GST_ELEMENT (sink_q));
  gst_pad_link (gst_element_get_pad (last, "src"),
      gst_element_get_pad (sink_q, "sink"));

  sink_thr = GST_ELEMENT (gst_thread_new ("sink_thread"));
  g_return_val_if_fail (sink_thr != NULL, NULL);

  sink = fakesink ();
  g_return_val_if_fail (sink != NULL, NULL);
  gst_bin_add (GST_BIN (sink_thr), GST_ELEMENT (sink));

  gst_bin_add (GST_BIN (pipeline), GST_ELEMENT (sink_thr));

  gst_pad_link (gst_element_get_pad (sink_q, "src"),
      gst_element_get_pad (sink, "sink"));

  return pipeline;
}

struct test
{
  char *name;
  char *params;
  GstPipeline *(*func) (int argc, int argi, char *argv[]);
};

static struct test tests[] = {
  {"simple", "ident_count [scheduler_name]", simple},
  {"queue", "ident_count [scheduler_name]", queue},
  {NULL, NULL, NULL}
};

int
main (int argc, char *argv[])
{
  GstPipeline *pipeline;
  int i;
  char *name;

  gst_init (&argc, &argv);

  if (argc < 3) {
    fprintf (stderr,
	"usage: %s iterations print_del mhz test_name [test_params...]\n",
	argv[0]);
    for (i = 0; tests[i].name; i++) {
      fprintf (stderr, "  %s %s\n", tests[i].name, tests[i].params);
    }
    exit (1);
  } else {
    iterations = atoi (argv[1]);
    print_del = atoi (argv[2]);
    mhz = atoi (argv[3]);
    name = argv[4];
  }

  pipeline = NULL;
  for (i = 0; tests[i].name && !pipeline; i++) {
    if (!strcmp (name, tests[i].name)) {
      pipeline = tests[i].func (argc, 5, argv);
    }
  }
  g_return_val_if_fail (pipeline != NULL, -1);

  /*xmlSaveFile("lat.gst", gst_xml_write(GST_ELEMENT(pipeline))); */

  gst_element_set_state (GST_ELEMENT (pipeline), GST_STATE_PLAYING);

  while (count < iterations) {
    gst_bin_iterate (GST_BIN (pipeline));
  }
  g_print ("\n");

  return 0;
}
